server/pulp/server/webservices/views/roles.py

```diff
@@ -0,0 +1,229 @@
+from pulp.server import exceptions as pulp_exceptions
+from pulp.server.auth import authorization
+from pulp.server.managers import factory
+from pulp.server.webservices.controllers.base import json_encoder as pulp_json_encoder
+from pulp.server.webservices.controllers.decorators import auth_required
+from pulp.server.webservices.views.util import (generate_json_response,
+                                                generate_json_response_with_pulp_encoder)
+
+
+class RolesView(View):
+    """
+    Views for roles.
+    """
+
+    @auth_required(authorization.READ)
+    def get(self, request):
+        """
+        List all roles.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :return: Response containing a list of roles
+        :rtype: django.http.HttpResponse
+        """
+        role_query_manager = factory.role_query_manager()
+        user_query_manager = factory.user_query_manager()
+        permissions_manager = factory.permission_manager()
+        roles = role_query_manager.find_all()
+        for role in roles:
+            role['users'] = [u['login'] for u in
+                             user_query_manager.find_users_belonging_to_role(role['id'])]
+
+            resource_permission = {}
+            # isolate schema change
+            if role['permissions']:
+                for item in role['permissions']:
+                    resource = item['resource']
+                    operations = item.get('permission', [])
+                    resource_permission[resource] = [permissions_manager.operation_value_to_name(o)
+                                                     for o in operations]
+
+            role['permissions'] = resource_permission
+        for role in roles:
+            link = {"_href": reverse('role_resource',
+                    kwargs={'role_id': role['id']})}
+            role.update(link)
+        return generate_json_response_with_pulp_encoder(roles)
+
+    @auth_required(authorization.CREATE)
+    def post(self, request):
+        """
+        Create a new role.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :return: Response containing the role
+        :rtype: HttpResponseRedirect
+        """
+        role_data = request.body_as_json
+        role_id = role_data.get('role_id', None)
+        display_name = role_data.get('display_name', None)
+        description = role_data.get('description', None)
+        manager = factory.role_manager()
+        role = manager.create_role(role_id, display_name, description)
+        link = {"_href": reverse('role_resource',
+                kwargs={'role_id': role['id']})}
+        role.update(link)
+        response = generate_json_response_with_pulp_encoder(role)
+        response['Location'] = iri_to_uri(link['_href'])
+        response.status_code= 201
+        response.reason_phrase= 'CREATED'
```

Comment:
Ok, I should have kept reading before commenting on the line above. I'm surprised django doesn't default this to the value in ``httplib.responses``. It's worth making sure; you might find that django does use that as a default. If it doesn't, it would be a great RFE and contribution to django, but you should in the mean time do this:

```
response.status_code = httplib.CREATED
response.reason_phrase = httplib.responses[response.status_code]
```